src/util/antlr/ANTLRErrorHandling.h

```diff
@@ -21,14 +22,40 @@ ExceptionMetadata generateMetadata(antlr4::Recognizer* recognizer,
 
 ExceptionMetadata generateMetadata(antlr4::ParserRuleContext* ctx);
 
-/**
- * antlr::ANTLRErrorListener that raises encountered syntaxErrors as
- * ParseException. The line, position in the line and antlr error message are
- * included as exception cause.
- */
+/*
+antlr::ANTLRErrorListener that raises encountered syntaxErrors as
+exceptions of type `GrammarParseException`. The line, position in the line and
+antlr error message are included as exception cause.
+
+For an example of a valid `GrammarParseException` see
+`InvalidSparqlQueryException`.
+*/
+template <typename GrammarParseException>
+requires std::derived_from<GrammarParseException, ParseException> &&
+         std::constructible_from<GrammarParseException, std::string_view,
+                                 std::optional<ExceptionMetadata>>
 struct ThrowingErrorListener : public antlr4::BaseErrorListener {
   void syntaxError(antlr4::Recognizer* recognizer,
                    antlr4::Token* offendingSymbol, size_t line,
                    size_t charPositionInLine, const std::string& msg,
-                   std::exception_ptr e) override;
+                   [[maybe_unused]] std::exception_ptr e) override {
+    throw GrammarParseException{generateExceptionMessage(offendingSymbol, msg),
+                                generateMetadata(recognizer, offendingSymbol,
+                                                 line, charPositionInLine)};
+  }
+
+ private:
+  std::string generateExceptionMessage(antlr4::Token* offendingSymbol,
+                                       const std::string& msg) {
+    if (!offendingSymbol) {
+      return msg;
+    } else if (offendingSymbol->getStartIndex() ==
+               offendingSymbol->getStopIndex() + 1) {
+      // This can only happen at the end of the query when a token is expected,
+      // but none is found. The offending token is then empty.
+      return absl::StrCat("Unexpected end of Query: ", msg);
+    } else {
+      return absl::StrCat("Token \"", offendingSymbol->getText(), "\": ", msg);
+    }
+  }
```

Comment:
This function can still be static and therefore be defined in the .cpp file (needn't even be part of the class, just make it a free function).
Also it shouldn't say "Unexcepted end of Query" (we don't necessarily parse a query)
but `"Unexpected end of input"`.

Reply:
Done, see [f78ab1b](https://github.com/ad-freiburg/qlever/pull/1044/commits/f78ab1bf8591952562c1ae3f5eb4b77c14dd5805). I also got rid of all the helper functions, that should have had their own util file.